src/client/pfs.go

```diff
@@ -438,6 +442,114 @@ func (c APIClient) ListBlock() ([]*pfs.BlockInfo, error) {
 	return blockInfos.BlockInfo, nil
 }
 
+// PutObject puts a value into the object store and tags it with 0 or more tags.
+func (c APIClient) PutObject(r io.Reader, tags ...string) (object *pfs.Object, retErr error) {
+	w, err := c.newPutObjectWriteCloser(tags...)
+	if err != nil {
+		return nil, sanitizeErr(err)
+	}
+	defer func() {
+		if err := w.Close(); err != nil && retErr == nil {
+			retErr = sanitizeErr(err)
+		}
+		if retErr == nil {
+			object = w.object
+		}
+	}()
+	if _, err := io.CopyBuffer(w, r, make([]byte, MaxMsgSize/2)); err != nil {
+		return nil, sanitizeErr(err)
+	}
+	// return value set by deferred function
+	return nil, nil
+}
+
+// GetObject gets an object out of the object store by hash.
+func (c APIClient) GetObject(hash string, writer io.Writer) error {
+	getObjectClient, err := c.ObjectAPIClient.GetObject(
+		c.ctx(),
+		&pfs.Object{Hash: hash},
+	)
+	if err != nil {
+		return sanitizeErr(err)
+	}
+	if err := grpcutil.WriteFromStreamingBytesClient(getObjectClient, writer); err != nil {
+		return sanitizeErr(err)
+	}
+	return nil
+}
+
+// ReadObject gets an object by hash and returns it directly as []byte.
+func (c APIClient) ReadObject(hash string) ([]byte, error) {
+	var buffer bytes.Buffer
+	if err := c.GetObject(hash, &buffer); err != nil {
+		return nil, err
+	}
+	return buffer.Bytes(), nil
+}
+
+// TagObject applies a tag to an existing object.
+func (c APIClient) TagObject(hash string, tags ...string) error {
+	var _tags []*pfs.Tag
+	for _, tag := range tags {
+		_tags = append(_tags, &pfs.Tag{Name: tag})
+	}
+	if _, err := c.ObjectAPIClient.TagObject(
+		c.ctx(),
+		&pfs.TagObjectRequest{
+			Object: &pfs.Object{Hash: hash},
+			Tags:   _tags,
+		},
+	); err != nil {
+		return sanitizeErr(err)
+	}
+	return nil
+}
+
+// InspectObject returns info about an Object.
+func (c APIClient) InspectObject(hash string) (*pfs.ObjectInfo, error) {
+	value, err := c.ObjectAPIClient.InspectObject(
+		c.ctx(),
+		&pfs.Object{Hash: hash},
+	)
+	if err != nil {
+		return nil, sanitizeErr(err)
+	}
+	return value, nil
+}
+
+// GetTag gets an object out of the object store by tag.
+func (c APIClient) GetTag(tag string, writer io.Writer) error {
+	getTagClient, err := c.ObjectAPIClient.GetTag(
+		c.ctx(),
+		&pfs.Tag{Name: tag},
+	)
+	if err != nil {
+		return sanitizeErr(err)
+	}
+	if err := grpcutil.WriteFromStreamingBytesClient(getTagClient, writer); err != nil {
+		return sanitizeErr(err)
+	}
+	return nil
+}
+
+// ReadTag gets an object by tag and returns it directly as []byte.
+func (c APIClient) ReadTag(tag string) ([]byte, error) {
+	var buffer bytes.Buffer
+	if err := c.GetTag(tag, &buffer); err != nil {
+		return nil, err
+	}
+	return buffer.Bytes(), nil
+}
+
+// Compact forces compaction of objects.
```

Comment:
There is not, and actually compaction doesn't really work with the caching layer at all here because it assumes immutability and to do compaction you need to move things around. I think I might rip out the compaction code before I merge this actually. However, I was thinking an optimization which would actually work is to make it so that you can write multiple objects in a single streaming request and on the back end those could easily be compacted into a single block. It seems like that would be pretty good for the case where there are several small files in the same commit. However in the interest of getting this merged into master so you guys can use it for the new pps stuff I think it's probably better if I do that as a separate PR.